### DISCUSSION TITLE
Nixie controller -- all Sequent hats compatible?

### DISCUSSION THREAD

**Discussion by Telemachus1977** [tagyoureit/nodejs-poolController#587] [Nixie]
My buddy is collecting the components for his nixie controller.  He has a very basic system..  dual body with a Pentair intelliflo-VS, two actuators and two lights.

He ordered this sequent relay board because it would get to him quickly and is cheap:

https://sequentmicrosystems.com/products/four-relays-four-inputs-for-raspberry-pi

Will REM access this ok through i2c?  What problems will he face compared to using the "recommended" automation and relay hats?  

I don't think the digital inputs work will with temperature sensors... He needs ADC, correct?

Perhaps this:

Check this out on Amazon
NOYITO 4-Channel 16-Bit ADC with PGA for RPI Raspberry PI 16 Bits I2C ADS1115 Module ADC 4 Channel for Raspberry Pi 3 2 Model B B+ https://a.co/d/c42xsEl

**Comment by tagyoureit** [comment]
We don't have definitions for either of these.  Could they be made when we have time, yes, but he couldn't use them now.  But honestly, spend the extra money and go with the recommended hardware.  Temperature sensors are one example where you would want other input types, like you say.  You could also check out how Hector is doing (https://github.com/tagyoureit/nodejs-poolController/discussions/416) and getting cheaper hardware has an additional cost of more wiring and configuration down the road.  Your buddy is already up to $61 when he could just spend $90 for the Mega-BAS and everything would work.

**Reply by arrmo** [reply]
Yes, agreed on the Mega-BAS and relays! The question is if they will plug on to the board. RPi wasn't available in any reasonable timeframe (only here a few days, RPi a couple weeks 😞) , and seemed to be ~ $200, vs $45 for this (drop in?) board. All the hats and such - yep, plan is to go with that exactly.

Thanks!

**Reply by Telemachus1977** [reply]
The nixie page the wiki implies that any ads1115 board should work.  Won't it be read the same through the i2c?

I have to admit, the board he got looks pretty tempting. RDS 485 built in, with four relays for only 45 bucks...  Some of those ads1115 boards are as cheap as 7 to $8.

**Comment by Telemachus1977** [comment]
The $15 ADS115 board works great with existing definitions in REM.

The 4 relay RS-485 board has all 4 relays functioning if it is selected as a relay board.  It has RS-485 working if selected as an io board.

@rstrouse If I were to create a new board definition and copy the rs-485 definitions from the io boards to one of the relay boards is that likely to work after a rebuild?

We aren't really looking at adding the input ports at this point since there is no ADC.

This cheap board looks like it is a great cheap answer to someone who just has a few components to run.

https://sequentmicrosystems.com/pages/four-relays-four-inputs-downloads

**Reply by tagyoureit** [reply]
You'll want to create a new definition in 2 places:
1. https://github.com/rstrouse/relayEquipmentManager/blob/master/i2c-bus/SequentIO.ts - This defines how REM communicates with the boards.  There are inherited classes for the different boards.
2. https://github.com/rstrouse/relayEquipmentManager/blob/master/devices/SequentIO.json - This is the board definition file.  It defines the inputs/outputs of the board as well as the design of the webpage (which has a LOT of nesting and can be a little bit confusing, but copy/paste carefully).  

To make a new device, these are the most important pieces:
```
      "id": 901,
      "enabled": true,
      "name": "Sequent MEGA-IND",
      "deviceClass": "SequentMegaIND",
      "module": "./SequentIO",
```
You should use id 903 and give a new device name/class.  You should also check out the multi-relay boards (https://github.com/rstrouse/relayEquipmentManager/blob/master/i2c-bus/I2c-Relay.ts) as they are a bit simpler, but since this has both input/output it should be in the 2 files mentioned above.

**Reply by Telemachus1977** [reply]
Yes just adding the relay function to one of the io boards and redefining it may be the easiest.

Does the "module" definition reference an external config file?  I noticed there are node red config files on sequent website.  Are any of these required, or is everything set up right in the two files that you referenced?

**Reply by tagyoureit** [reply]
We don't use any node-red.  Everything you need should already be included.  The module should stay the same.  It is the base of the typescript file `SequentIO.ts` which contains the deviceClass in that ts file (which you should copy and rename).

**Reply by Telemachus1977** [reply]
A little bit of cut and pasting and we are in business.  My buddy is controlling his pump via RS485 and relays are working.  Thanks for the tips!

At some point we might ask you gentlemen to tidy up our hack job to include in future releases.  This cheap board would be great for someone with a pump and less than 4 circuits.

We threw out the input tab entirely since no onboard ADC...  But I guess there could be some potential in the future to be able to read inputs up to 240 volts.

![SmartSelect_20220906_163928_Message+](https://user-images.githubusercontent.com/25725743/188759385-347b0637-b4e2-4a53-ad2f-88f03c753680.jpg)
![SmartSelect_20220906_163901_Message+](https://user-images.githubusercontent.com/25725743/188759390-784a3da4-3543-4219-94b8-d85526510bdc.jpg)

**Reply by tagyoureit** [reply]
Awesome!  Do you know how to do a pull request?  We can include this code.  Or if you want to copy/paste the source files here we can manually integrate them.  Glad you got it up and running.

**Reply by Telemachus1977** [reply]
I spoke too soon.  Relays aren't feeding through to the I2c TS files yet.

Can I have ID 903 call for both ./Sequintio.ts AND ./I2C-Relay.ts?

It would be cleaner than copying all the code over from one to the other like I did.

Ie:

      "id": 903,
      "enabled": true,
      "name": "Sequent 4 Relay 4 input",
      "deviceClass": "Sequent4Rel4in",
      "module": "./SequentIO",
       "module": "./I2c-Relay",

**Reply by tagyoureit** [reply]
Yup, that's fine.

**Reply by rstrouse** [reply]
I think that @tagyoureit misunderstood what you are trying to do.  Typescript does not support polymorphism so you cannot create one class from two siblings.  This will not work and you will likely have a json error if you try to parse the above snippet.  What you can do however, is put relay code into a 4 relay/input class in the SequentIO.ts file.  This would be greatly slimmed down code since it does not have to code for the differences between all of the other relay drivers.

If you would like some help there I can assist in getting that going.

**Reply by tagyoureit** [reply]
Yes - listen to @rstrouse.  Either I did read all of the words in your post but only a subset made it to my brain, OR you edited your post after I replied.  :). Either way, you can't combine the modules like that.  "Copy you must" - Yoda.

**Reply by Telemachus1977** [reply]
Yeah it didn't like calling for two modules at all.

What I ended up doing was creating a brute force TS file that combined everything from SequentIO and I2c-Relay.

I couldn't get any of the info to show up on the General tab of the MultiIO json files (FW, Voltage, temp, etc) so I went the other direction and took the Relay file and added in the RS-485 info.

As you stated, it is pretty beefy with a lot of relay code that is unnecessary for this specific board.  However, it does give some fun things to play with as the eight relay boards turn the optically isolated inputs on and off.   If only the optoisolators could work in reverse, these could actually be pretty useful.  For now using these files the inputs are just on all of the time.

I'm no coder, and nothing about this is elegant, but it is working.  RS-485 and 4 relays for $45 is a nice starter kit for REM IMHO.

@rstrouse I uploaded the files on the REM code for you to tighty up at some point.  Eventually I would like to figure out how to get the inputs to actually do something useful.

**Comment by Telemachus1977** [comment]
![Screenshot_20220907_123941_VNC Viewer](https://user-images.githubusercontent.com/25725743/188982703-466fba4d-d37b-44fa-af55-7df82d91bd62.jpg)

**Reply by rstrouse** [reply]
Where did you post up the code?

**Reply by Telemachus1977** [reply]
In uploaded it to Telemachus1977/relayEquipmentManager

The files are under the code as patches.

**Reply by Telemachus1977** [reply]
We were able to verify tonight that this same code works with the Sequent 8 relay industrial board which also includes RS-485.  This is more useful than the 4 input 4 relay board for the same price.

https://sequentmicrosystems.com/products/8-relays-stackable-card-for-raspberry-pi

$45 gives you 8 relays and RS-485
$12 gives you a stackable ADS1115 board with 4 inputs

You lose the ability to power the Raspberry Pi off 24V but there are many solutions for that that still put the total price at less than a MEGA-BAS.  I suppose you also lose watchdog ability... But that can also be added if needed.


Digging around on sequent's site there is also a 16 relay board with RS-485.

I feel like the best solution for the TS files would be to add the RS-485 config to the standard I2c-Relay file as several relay boards support it.

The RS-485 tab should pop up in the relay.json file for the relays that support it or it could always be there with a note that only some boards support it.

**Comment by rstrouse** [comment]
Ok so I added support for the 4 relays and 4 inputs if you pull REM.  The relay interface was added to the Sequent functions and the opto-isolated digital inputs are also supported on the card.

There are a couple of things about this board you should know.
1. There is no software control of the RS485 port.  This is because it does not support MODBUS but we don't use that anyway.  To enable the RS485 you must use the jumper on the board.
2. The digital inputs are only on and off on that board.  They are opto-isolated single bit sensors and do not convert analog signals.  The trigger is > 3v or > 48v depending on the connector used.
3. The digital inputs are not dry contact sensors and will need external power to operate as such.  They are however optically isolated so an external power source can be used to detect power on for peripheral devices.

The 16 relay board has one major drawback.  The relays are only Normally Open.  This means that there is no ability to control valves with it.  I'm sure they did this because of the limits regarding the connector space.

**Reply by Telemachus1977** [reply]
Thanks.   You make this look so easy lol.

The 8 relay board is truly the better option here as I don't see a meaningful common use for the isolated inputs.  Can you also add the sequent 8 IND board to the I/O with the RS-485 tab added?

The 16 relay board has NC on one side and NO on the other.   Technically couldnt a valve still be used but would require two relays used in tandom such that as one closes another opens?  They share commons which makes the wiring no more burdensome. As lights and heaters don't need both there is still a place for it.   There is a 2.5A limitation per channel, which could be overcome by some incandescent light bulbs.   The 8 channel industrial relay is limited to 4 amps, which could still be stretched by some bulbs.

![Screenshot_20220908_153417_Chrome](https://user-images.githubusercontent.com/25725743/189238769-c547a3a4-d0a1-4de5-b210-e56ac5dd6f5a.jpg)

**Reply by rstrouse** [reply]
I don't know that I would switch 24vdc power with that card.  While the relay says it can take it I suspect that it would get beat up with flyback from the contacts.  

Using 2 relays is a possibility but why when you can ensure a mechanical switch so that both legs of a valve are never energized.  Switch timing would become too important.

This actually has more utility
https://sequentmicrosystems.com/collections/all-io-cards/products/three-40a-240v-relays-rs485-for-raspberry-pi

**Reply by Telemachus1977** [reply]
Agreed.   That guy is a beast!

The beaglebone adapter card is also pretty cool. 

https://sequentmicrosystems.com/collections/accessories/products/beagle-pi-raspberry-pi-emulator-using-beagle-board

I'll bet the beaglebone black is a powerful as the raspberry pi 3b+  I'm using. 

Thanks for all your help with this. 

Can you also add the sequent 8 IND board to the I/O with the RS-485 tab added?

**Reply by rstrouse** [reply]
There is also no modbus control on the Sequent 8 IND board.  There are no settings that can be set with that.  You have to use the J2 jumper block and jump the R pins.  njsPC will control the baud rate and such from there.  

The reason why there is an RS485 tab on the BAS and IND card is that these both support modbus that has to be disabled on the 485 controller.

**Reply by Telemachus1977** [reply]
That makes sense.   

 So there is nothing in the JSON file nor the TS file that are required to make RS-485 work?  Just place the jumper and the system will pick it up?

**Reply by rstrouse** [reply]
Yessir.  No software config on the card itself.

**Reply by Telemachus1977** [reply]
Well I spent an illuminating day learning a little bit of javascript and typescript for nothing, lol.

**Reply by rstrouse** [reply]
Nuttin wrong with that.  Picking up something new is the best part of all this.

**Reply by Telemachus1977** [reply]
I feel like I learned something through every interaction with you. Appreciate all you have done for this.

**Reply by rstrouse** [reply]
Hard refresh. CTRL + F5.  That selection will disappear.  It is not required for this relay setup.

**Reply by Telemachus1977** [reply]
For some reason I thought the controller let you select between modbus and RS485. I figured it out and edited my question hoping you wouldn't see it.

**Reply by Telemachus1977** [reply]
The next thing we will have to figure out is the sequent fan board. The one my buddy got came with a higher firmware than the ones shown in the examples on the wiki. It was unable to read the fan (board) temperature and inserting feeds for CPU temp from internal also did not work. None of the scripts seems to work other than being able to turn the fan on to a certain percentage
